lib/spatial_features/importers/file.rb

```diff
@@ -3,22 +3,40 @@
 module SpatialFeatures
   module Importers
     class File < SimpleDelegator
-      def initialize(data, *args)
+      INVALID_ARCHIVE = "Archive did not contain a .kml or .shp file. Supported formats are KMZ, KML, and zipped ArcGIS shapefiles.".freeze
+
+      def initialize(data, *args, **options)
+        # the current_file param is passed by `::create` after it has opened a zip
+        # archive and extracted the KML and SHP files
+        current_file = options.delete(:current_file)
+
         begin
-          file = Download.open(data, unzip: [/\.kml$/, /\.shp$/], downcase: true)
+          current_file ||= Download.open_each(data, unzip: [/\.kml$/, /\.shp$/], downcase: true).first
         rescue Unzip::PathNotFound
-          raise ImportError, "Archive did not contain a .kml or .shp file. Supported formats are KMZ, KML, and zipped ArcGIS shapefiles."
+          raise ImportError, INVALID_ARCHIVE
         end
 
-        case ::File.extname(file.path.downcase)
+        case ::File.extname(current_file.path.downcase)
         when '.kml'
-          __setobj__(KMLFile.new(file, *args))
+          __setobj__(KMLFile.new(current_file, *args))
         when '.shp'
-          __setobj__(Shapefile.new(file, *args))
+          __setobj__(
+            Shapefile.new(data, *args, **options, shp_file_name: ::File.basename(current_file&.path))
```

Comment:
Let's leave a comment as to why we're passing data since I think the natural inclination would be to want to pass the file we just unzipped. In this case it's now only performing a check to see if we _will_ be able to handle the file.